### PR TITLE
Add support for Chameleon

### PR DESCRIPTION
**Pull request by nopperl**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

This PR adds support for the [Chameleon](https://github.com/facebookresearch/chameleon) model. For now, this implementation only supports text->text inference and serves as base to implement the (more interesting) image->text, text->image and interleaved pipelines. However, such an implementation will probably require some changes to the CLI and internal architecture, so I suggest to do this in a separate PR.

Chameleon is based on the Llama-2 architecture with the following changes:
* different (pre-)tokenizer
* qk-norm
* swin-norm

Note 1: in order to enable text->text inference, the [image token logits are suppressed](https://github.com/ggerganov/llama.cpp/pull/8543/files#diff-57725fa4c99fac28d0283eaa8b59c5e030355af0c8fbaf33145a64e3f2aa3420R13845-R13852) similar to the [HF implementation](https://github.com/huggingface/transformers/blob/24cfcc2114b5daf6ca1cce6e838ee9aa17096e62/src/transformers/models/chameleon/modeling_chameleon.py#L1549-L1551). This needs to be removed when support for images is added.

Note 2: I implemented swin-norm, but I haven't tested it yet, as it is only used by Chameleon-30B.

To test it:

```
git clone https://huggingface.co/facebook/chameleon-7b
./convert-hf-to-gguf.py chameleon-7b
build/bin/llama-cli -m chameleon-7b/ggml-model-f16.gguf --temp 0.8 -s 1000 -n 50 -p "Language modeling is " -ngl 33
```

Output:

```
Language modeling is “the task of predicting the next word in a sequence of text, given the previous words.”

To implement a language model, we can use a neural network with a bidirectional LSTM layer and a softmax output layer.
```

Reference (requires transformers>=4.43.0.dev0):

```
from transformers import AutoModelForCausalLM, AutoTokenizer, set_seed
set_seed(1000)
model = AutoModelForCausalLM.from_pretrained("facebook/chameleon-7b", device_map="auto")
tokenizer = AutoTokenizer.from_pretrained("facebook/chameleon-7b")
prompt = "Language modeling is "
inputs = tokenizer.encode(prompt, return_pt=True)
out = model.generate(inputs, max_new_tokens=40)
tokenizer.decode(out)
```

Reference output:

```
Language modeling is “the task of predicting the next word in a sequence of text given the previous words.”

In other words, it's a machine learning model that takes a sequence of text as input
```

Partially addresses #7995.

### CODE REVIEW

**Comment by nopperl** [comment]
I have uploaded GGUFs to test this PR with [here](https://huggingface.co/nopperl/chameleon-7b-GGUF).

File: gguf-py/gguf/gguf_writer.py [hunk 1]
```diff
@@ -704,6 +704,9 @@ def add_middle_token_id(self, id: int) -> None:
     def add_eot_token_id(self, id: int) -> None:
         self.add_uint32(Keys.Tokenizer.EOT_ID, id)
 
+    def add_swin_norm(self, value: bool) -> None:
+        self.add_bool(Keys.LLM.SWIN_NORM.format(arch=self.arch), value)
+
```
Comment (compilade) [inline gguf-py/gguf/gguf_writer.py:709]: This could be moved between `add_expert_weights_scale` and `add_layer_norm_eps`.

Otherwise it feels a bit out of place near the tokenizer options.

File: src/llama.cpp [hunk 2]
```diff
@@ -15581,6 +15842,16 @@ struct llm_tokenizer_bpe {
                     "\\p{N}",
                 };
                 break;
+            case LLAMA_VOCAB_PRE_TYPE_CHAMELEON:
+                regex_exprs = {
+                    "<sentinel:[0-9]+>",  // Sentinel tokens
+                    "(IMGIMG)((A|B|C|D|E|F|G|H|I){1,4})Z",  // Image tokens
+                    "([\t\n]|    |  )",  // directly from tokenizer.json
+                    "\\p{N}", // Individual digits
+                    "[\\p{P}\\$\\+<=>\\^~\\|`]+",  // Punctuation
```
Comment (compilade) [inline src/llama.cpp:15851]: The proper regex for punctuation would need the `Po` category, which isn't supported until #8606. But for the ASCII-only part, HF `tokenizers` use Rust's [`is_ascii_punctuation`](https://doc.rust-lang.org/std/primitive.char.html#method.is_ascii_punctuation).

And I *think* `Isolated` means consecutive punctuation should be split (so no trailing `+`).

```suggestion
                    "[\\p{P}!-/:-@\\[-`{-~]",  // Punctuation, Isolated
```

(using ranges from the Rust documentation for conciseness and for completeness)

File: src/llama.cpp [hunk 3]
```diff
@@ -15581,6 +15842,16 @@ struct llm_tokenizer_bpe {
                     "\\p{N}",
                 };
                 break;
+            case LLAMA_VOCAB_PRE_TYPE_CHAMELEON:
+                regex_exprs = {
+                    "<sentinel:[0-9]+>",  // Sentinel tokens
+                    "(IMGIMG)((A|B|C|D|E|F|G|H|I){1,4})Z",  // Image tokens
+                    "([\t\n]|    |  )",  // directly from tokenizer.json
```
Comment (compilade) [inline src/llama.cpp:15849]:
```suggestion
                    "([\\t\\n]|    |  )",  // directly from tokenizer.json
```

But I think the behavior of the regex would be the same anyway, so feel free to ignore this.

File: src/llama.cpp [hunk 4]
```diff
@@ -15581,6 +15842,16 @@ struct llm_tokenizer_bpe {
                     "\\p{N}",
                 };
                 break;
+            case LLAMA_VOCAB_PRE_TYPE_CHAMELEON:
+                regex_exprs = {
+                    "<sentinel:[0-9]+>",  // Sentinel tokens
+                    "(IMGIMG)((A|B|C|D|E|F|G|H|I){1,4})Z",  // Image tokens
```
Comment (compilade) [inline src/llama.cpp:15848]: Side note: special tokens don't usually need to be considered by the split regex (since HF `tokenizers` pre-tokenize them specially anyway, and `llama.cpp` too in `tokenizer_st_partition`), but it seems like the upstream tokenizer uses these regular expressions in the `pre_tokenzier`, so they likely did not know this. Keeping this is appropriate, especially since their sentinel token regex matches more than only the sentinel tokens in the vocab.

Other note: this is the model with *the biggest number* of special tokens I have ever seen. 16384 special tokens!?

Comment (nopperl) [reply src/llama.cpp:15848]: >special tokens don't usually need to be considered by the split regex (since HF tokenizers pre-tokenize them specially anyway, and llama.cpp too in tokenizer_st_partition)

Good to know! Unsure if they are even intentionally used, but probably best to keep them. I'll add a comment.

>Other note: this is the model with the biggest number of special tokens I have ever seen. 16384 special tokens!? That's like 1/4 of its vocab!
true  😄  A big part of them are the image tokens from the VQVAE codebook (8192 tokens).

**Comment by nate-lrt** [comment]
will this ever get added :(

**Comment by nopperl** [comment]
I think it would still be a good addition. I've resolved all conflicts with master now, so it should be ready to merge.

**Review by ggerganov — APPROVED** [review]

**Comment by arch-btw** [comment]
Thank you @nopperl looks like it got merged!

**Comment by MasterScrat** [comment]
@nopperl any plans to tackle image->text and text->image?

**Comment by nopperl** [comment]
@MasterScrat currently no plans, sorry for the late reply. AFAIK multimodal support would require a refactor of llama.cpp (https://github.com/ggerganov/llama.cpp/issues/8010#issuecomment-2376339571). I'd love to work on it, but don't have the time right now.
